src/main/java/org/sonatype/nexus/repository/composer/internal/ComposerJsonProcessor.java

```diff
@@ -239,6 +243,94 @@ public Content buildProviderJson(final Repository repository,
         ContentTypes.APPLICATION_JSON));
   }
 
+  /**
+   * Merges an incoming set of packages.json files.
+   */
+  public Content mergePackagesJson(final Repository repository, final List<Payload> payloads) throws IOException {
+    Set<String> names = new HashSet<>();
+    for (Payload payload : payloads) {
+      Map<String, Object> json = parseJson(payload);
+      Map<String, Object> providers = (Map<String, Object>) json.get(PROVIDERS_KEY);
+      names.addAll(providers.keySet());
+    }
+    return buildPackagesJson(repository, names);
+  }
+
+  /**
+   * Merges incoming provider JSON files, producing a merged file containing only the minimal subset of fields that we
+   * need to download artifacts.
+   */
+  public Content mergeProviderJson(final Repository repository, final List<Payload> payloads, final DateTime now)
+      throws IOException
+  {
+    String time = now.withZone(DateTimeZone.UTC).toString(timeFormatter);
+
+    // TODO: Make this more robust, right now it makes a lot of assumptions and doesn't deal with bad things well,
+    // can probably consolidate this with the handling for rewrites for proxy (or at least make it more rational).
```

Comment:
Suggestion: If possible I would consider breaking it down so it would be more manageable for someone else to visit

Reply:
I would at least break out the individual concerns i.e. buildDistInfo, buildPackageInfo

Reply:
> I would at least break out the individual concerns i.e. buildDistInfo, buildPackageInfo

Done, also consolidated some near-duplicate code in https://github.com/sonatype-nexus-community/nexus-repository-composer/pull/14/commits/5ae09130b771886d44e76225f04a88dcd86965cf.